lasagne/utils.py

```diff
@@ -321,14 +322,31 @@ def create_param(spec, shape, name=None):
     This function is called by :meth:`Layer.add_param()` in the constructor
     of most :class:`Layer` subclasses. This enables those layers to
     support initialization with numpy arrays, existing Theano shared variables
-    or expressions, and callables for generating initial parameter values.
+    or expressions, and callables for generating initial parameter values,
+    Theano expressions, or shared variables.
     """
     shape = tuple(shape)  # convert to tuple if needed
     if any(d <= 0 for d in shape):
         raise ValueError((
             "Cannot create param with a non-positive shape dimension. "
             "Tried to create param with shape=%r, name=%r") % (shape, name))
 
+    err_prefix = "cannot initialize parameter %s: " % name
+    if hasattr(spec, '__call__'):
+        spec = spec(shape)
+        err_prefix += "the %s returned by the provided callable"
+    else:
+        err_prefix += "the provided %s"
+
+    if np.isscalar(spec):
+        spec = floatX(spec)
+
+    if isinstance(spec, np.ndarray):
+        if spec.shape != shape:
+            raise ValueError("%s has shape %s, should be %s" %
+                             (err_prefix % "numpy array", spec.shape, shape))
+        spec = theano.shared(floatX(spec))
```

Comment:
Ah, you sneaked in a `floatX` here. This wasn't present before in the codepath for an ndarray. That's why the doctest for the EmbeddingLayer fails. We should have a more direct test that checks whether the dtype of an ndarray is retained in `create_param`!